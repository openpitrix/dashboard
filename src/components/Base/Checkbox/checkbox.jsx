import React, { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Icon from '../Icon';
import styles from './index.scss';

export default class Checkbox extends Component {
  static propTypes = {
    checked: PropTypes.bool,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func,
    value: PropTypes.string
  };

  static defaultProps = {
    className: '',
    disabled: false,
    checked: false
  };

  constructor(props) {
    super(props);

    this.state = {
      isChecked: props.checked
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.state.isChecked) {
      this.setState({ isChecked: nextProps.checked });
    }
  }

  handleToggleCheck = e => {
    const { disabled, onChange } = this.props;

    if (disabled) {
      return;
    }
    onChange && onChange(e);
  };

  render() {
    const {
      style,
      className,
      disabled,
      children,
      isFold,
      fold,
      value
    } = this.props;
    const { isChecked } = this.state;
    const labelClass = isFold ? styles.foldCheckbox : styles.checkbox;

    return (
      <label
        className={classNames(labelClass, className, {
          [styles.checked]: isChecked
        })}
        disabled={disabled}
        style={style}
      >
        {isChecked && <Icon name="check" />}
        {isFold && fold && <Icon name="check-fold" size={28} />}
        {isFold && !fold && <Icon name="check-unfold" size={28} />}
        <input
          type="checkbox"
          value={value}
          checked={isChecked}
          onChange={this.handleToggleCheck}
        />
        {children}
      </label>
    );
  }
}
