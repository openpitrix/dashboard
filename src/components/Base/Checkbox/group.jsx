import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { remove } from 'lodash';

import styles from './index.scss';

export default class CheckboxGroup extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.arrayOf(PropTypes.node)
    ]),
    className: PropTypes.string,
    name: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.array
  };

  static defaultProps = {
    children: [],
    values: [],
    name: ''
  };

  handleChange = e => {
    e.stopPropagation();

    const { name, onChange } = this.props;
    const targetValue = e.target.value;
    const { values } = this.props;

    if (e.target.checked) {
      values.push(targetValue);
    } else {
      remove(values, v => v === targetValue);
    }
    onChange && onChange(values, name);
  };

  render() {
    const {
      className, name, children, values
    } = this.props;

    const childNodes = React.Children.map(children, child => React.cloneElement(child, {
      ...child.props,
      key: `check-${child.props.value}`,
      name: child.props.name || name,
      checked: values.indexOf(child.props.value) !== -1,
      onChange: this.handleChange
    }));

    return (
      <div className={classNames(styles.group, className)}>{childNodes}</div>
    );
  }
}
