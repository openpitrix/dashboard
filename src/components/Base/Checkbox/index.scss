@import '~scss/vars';

.group {
  display: inline-block;
  .checkbox {
    margin-right: 23px;
  }
}

.checkbox {
  position: relative;
  display: inline-block;
  padding-left: 20px;
  transition: all 0.1s cubic-bezier(0.71, -0.46, 0.88, 0.6);

  input[type="checkbox"] {
    width: 0;
    height: 0;
    opacity: 0;
    user-select: none;
    outline: none;
  }

  :global {
    .icon {
      display: none;
      position: absolute;
      left: 1px;
      top: 50%;
      font-size: $size-small;
      color: #fff;
      transform: translate(0 ,-50%) scale(0.7);
      cursor: pointer;
    }
  }

  &::before {
    content: "";
    position: absolute;
    left: 1px;
    top: 50%;
    z-index: 2;
    width: 14px;
    height: 14px;
    font-style: normal;
    background-color: $background-color-lighter;
    border: 1px solid $border-lighter;
    border-radius: 2px;
    box-shadow: 0 2px 4px 0 rgba(48, 14, 86, 0.04);
    box-sizing: border-box;
    transform: translate(0 ,-50%);
    cursor: pointer;
  }

  &.checked {
    &::before {
      //border-radius: 2px;
      border-color: $P75;
      background-color: $P75;
    }
    :global {
      .icon {
        display: block;
        z-index: 3;
      }
    }
  }

  &.disabled {
    &::before {
      background-color: $background-color;
      cursor: not-allowed;
    }
  }
}
