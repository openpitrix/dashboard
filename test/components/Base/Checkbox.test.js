import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Checkbox from 'components/Base/Checkbox';

describe('Base/Checkbox', () => {
  it('basic render', () => {
    const wrapper = render(
      <Checkbox className="test" name="test" disabled>
        option1
      </Checkbox>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('group render', () => {
    const wrapper = render(
      <Checkbox.Group name="test" values={['1', '3']}>
        <Checkbox value="1">option1</Checkbox>
        <Checkbox value="2">option2</Checkbox>
        <Checkbox value="3">option3</Checkbox>
      </Checkbox.Group>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('call onChange', () => {
    const mockChange = jest.fn();
    const wrapper = mount(<Checkbox onChange={mockChange} checked />);
    wrapper.find('input').simulate('change');

    const { isChecked } = wrapper.state();

    expect(isChecked).toBeTruthy();
    expect(mockChange).toHaveBeenCalled();
    expect(mockChange).toHaveBeenCalledTimes(1);
    expect(mockChange.mock.calls[0][0].target.checked).toEqual(isChecked);
  });
});
