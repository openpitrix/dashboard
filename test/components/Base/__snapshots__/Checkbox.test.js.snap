// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Checkbox basic render 1`] = `
<label
  class="checkbox test"
  disabled=""
>
  <input
    type="checkbox"
  />
  option1
</label>
`;

exports[`Base/Checkbox group render 1`] = `
<div
  class="group"
>
  <label
    class="checkbox checked"
  >
    <span
      class="icon icon-check"
      style="width:16px;height:16px"
    >
      <svg
        class="qicon qicon-check qicon-light"
        style="--primary-color:#fff;--secondary-color:#fff"
      >
        <use
          href="#check"
        />
      </svg>
    </span>
    <input
      checked=""
      type="checkbox"
    />
    option1
  </label>
  <label
    class="checkbox"
  >
    <input
      type="checkbox"
    />
    option2
  </label>
  <label
    class="checkbox checked"
  >
    <span
      class="icon icon-check"
      style="width:16px;height:16px"
    >
      <svg
        class="qicon qicon-check qicon-light"
        style="--primary-color:#fff;--secondary-color:#fff"
      >
        <use
          href="#check"
        />
      </svg>
    </span>
    <input
      checked=""
      type="checkbox"
    />
    option3
  </label>
</div>
`;
